File: nano-lightroom/nano_client.py
"""Thin wrapper around google-genai for Nano Banana image editing."""
from __future__ import annotations

import io
from dataclasses import dataclass

from PIL import Image
from google import genai
from google.genai import types

MODEL = "gemini-2.5-flash-image"


@dataclass
class EditResult:
    image: Image.Image | None
    text: str | None
    error: str | None = None


class NanoClient:
    def __init__(self, api_key: str):
        if not api_key:
            raise ValueError("Missing GEMINI_API_KEY")
        self.client = genai.Client(api_key=api_key)

    def edit(self, image: Image.Image, prompt: str) -> EditResult:
        """Send image + prompt to Nano Banana, return the edited image."""
        try:
            response = self.client.models.generate_content(
                model=MODEL,
                contents=[prompt, image],
                config=types.GenerateContentConfig(
                    response_modalities=["IMAGE", "TEXT"],
                ),
            )
        except Exception as exc:  # surface API errors to UI
            return EditResult(image=None, text=None, error=str(exc))

        out_image: Image.Image | None = None
        out_text: str | None = None

        candidates = getattr(response, "candidates", None) or []
        for cand in candidates:
            content = getattr(cand, "content", None)
            if not content:
                continue
            for part in getattr(content, "parts", []) or []:
                inline = getattr(part, "inline_data", None)
                if inline and getattr(inline, "data", None):
                    out_image = Image.open(io.BytesIO(inline.data))
                    out_image.load()
                elif getattr(part, "text", None):
                    out_text = (out_text or "") + part.text

        if out_image is None:
            return EditResult(
                image=None,
                text=out_text,
                error="No image returned (model may have refused or returned text only).",
            )
        return EditResult(image=out_image, text=out_text)

File: nano-lightroom-mobile/lib/nanoClient.js
// Direct REST call to the Gemini Nano Banana endpoint.
// We use fetch instead of @google/genai because the SDK depends on Node APIs
// that aren't available in React Native.

const MODEL = 'gemini-2.5-flash-image';
const ENDPOINT = `https://generativelanguage.googleapis.com/v1beta/models/${MODEL}:generateContent`;

/**
 * Edit a single image with Nano Banana.
 *
 * @param {string} apiKey
 * @param {string} prompt
 * @param {string} base64Image  raw base64 string (no data: prefix)
 * @param {string} mimeType     e.g. 'image/jpeg'
 * @returns {Promise<{ base64: string, mimeType: string, text: string | null }>}
 */
export async function editImage(apiKey, prompt, base64Image, mimeType = 'image/jpeg') {
  if (!apiKey) throw new Error('Missing Gemini API key');

  const body = {
    contents: [
      {
        parts: [
          { text: prompt },
          { inline_data: { mime_type: mimeType, data: base64Image } },
        ],
      },
    ],
    generationConfig: {
      responseModalities: ['IMAGE', 'TEXT'],
    },
  };

  const response = await fetch(`${ENDPOINT}?key=${apiKey}`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body),
  });

  if (!response.ok) {
    const errText = await response.text();
    throw new Error(`Gemini API ${response.status}: ${errText.slice(0, 300)}`);
  }

  const json = await response.json();
  const parts = json?.candidates?.[0]?.content?.parts ?? [];

  let imagePart = null;
  let text = null;
  for (const part of parts) {
    if (part.inline_data?.data || part.inlineData?.data) {
      imagePart = part.inline_data ?? part.inlineData;
    } else if (part.text) {
      text = (text ?? '') + part.text;
    }
  }

  if (!imagePart) {
    throw new Error(text ? `Model returned text only: ${text.slice(0, 200)}` : 'No image in response');
  }

  return {
    base64: imagePart.data,
    mimeType: imagePart.mime_type ?? imagePart.mimeType ?? 'image/png',
    text,
  };
}
